crates/wasmtime/src/runtime/vm/vmcontext.rs

```diff
@@ -1188,6 +1188,25 @@ pub struct VMStoreContext {
     /// situation while this field is read it'll never classify a fault as an
     /// guard page fault.
     pub async_guard_range: Range<*mut u8>,
+
+    /// The last Wasm exit to host was due to a trap.
+    ///
+    /// This is set whenever we update the exit state *from the host*
+    /// when handling a trap. It allows us to interpret the exit PC
+    /// correctly -- that is, either pointing *to* a trapping
+    /// instruction, or *after* a call (a single PC could be both
+    /// after a call and at a trapping instruction!).
+    ///
+    /// It is set *only* from host code, but is kept here alongside
+    /// the other last-exit state for consistency.
+    pub last_wasm_exit_was_trap: UnsafeCell<bool>,
+
+    /// The last Wasm exit to host, if via trap, had this FP in guest
+    /// code (*not* in a trampoline).
+    ///
+    /// Either this field or `last_wasm_exit_trampoline_fp` is set,
+    /// never both. This field is always set from the host.
+    pub last_wasm_exit_trap_fp: UnsafeCell<usize>,
```

Comment:
Personally I'm a bit wary about how this is an ever-expanding structure with a huge matrix of fields all customized for various purposes. Especially with a whole litany of `last_*` fields for all sorts of purposes it's kind of hard to keep it all straight.

Especially when I think these fields aren't actually modified from compiled code at all, would it be possible to store this elsewhere and/or infer it from control flow and/or other state? For example instead of `last_wasm_exit_trap_fp` could this store directly in `last_wasm_exit_trampoline_fp`? And instead of storing `last_wasm_exit_was_trap` could that be carried around as a paramter in relevant locations instead?

Reply:
One of my worries is that as this state grows and grows more for all sorts of one-off things it becomes more and more difficult to ensure that everything is always kept in sync with everything else. I find it's more robust to reuse existing state and/or minimize state where possible to ensure that everything is exercised more often, where possible at least

Reply:
Also, regarding storing elsewhere: because it's part of the "last activation" state, it needs to be saved/restored along with other parts of the `VMStoreContext` when, for example, `CallThreadState`s are pushed/popped on the activation chain when fibers suspend. That all works with immutable borrows today and directly accesses the unsafe cells in this location; that's why I kept it here and not somewhere else.

Reply:
I did the refactor to remove "trampoline FP" (in normal case) and "trap FP" (in trap case) bifurcation in favor of one FP in 7fd1f60e7c. We still need the `was_trap` state (and to save/restore it); there's no way to avoid that bit as far as I can tell.